consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,394 @@
+able to communicate across a network. The group should be able to tolerate `f`
+failures without losing liveness or safety.
+
+All participants know the other participants in the system and it is assumed
+that all messages are signed in some way such that participants in the protocol
+know which participant a message came from and that messages cannot be forged.
+
+Only messages from participants are considered.
+
+For messages containing a value only messages with valid values are considered.
+
+Participants are able to broadcast messages to all other participants. In the
+case that a participant is off-line or somehow inaccessible they will not
+receive broadcast messages and there is no mechanism for these messages to be
+re-sent.
+
+We refer to a consensus instance to mean consensus for a specific height,
+consensus instances are independent (they have no shared state).
+
+#### Instance state
+`Sc - current participant state`\
+`Hc - current height`\
+`Rc - current round`\
+`Rd - desired round`\
+`Vc - currently proposed value`\
+`PCc - current prepared certificate`\
+
+### Algorithm
+See supporting [functions](#Functions) and [notation](#Appendix-1-Notation).
+```
+upon: FC_E
+  Hc ← Hc+1
+  Rc ← 0
+  Rd ← 0
+  Sc ← AcceptRequest
+  Vc ← nil
+  schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)
+
+upon: <R_E, Hc, V> && Sc = AcceptRequest
+  if Rc = 0 && isProposer(Hc, Rd) {
+    bc(<PP_T, Hc, 0, V, nil>)
+  }
+
+upon: <PP_T, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
+  if (Rd > 0 && validRCC(Hc, Rd, V, RCC)) || (Rd = 0 && RCC = nil)  {
+    Rc ← Rd
+    Vc ← V
+    Sc ← Preprepared
+    bc(<P_T, Hc, Rd, Vc>)
+  }
+
+upon: M ← { <T, Hc, Rd, Vc> : T ∈ {P_T, C_T} } && |M| > 2f+1 && Sc ∈ {AcceptRequest, Preprepared} 
+  Sc ← Prepared
+  PCc ← <PC_T, M, Vc>
+  bc(<C_T, Hc, Rd, Vc>)
+
+upon: M ← { <C_T, Hc, Rd, Vc> } && |M| > 2f+1 && Sc ∈ {AcceptRequest, Preprepared, Prepared} 
+  Sc ← Committed
+  deliverValue(Vc)
+
+upon: m<RC_T, Hc , R, PC> && (PC = nil || validPC(PC)) 
+  if R < Rd {
+    send(<RC_T, Hc, Rd, PCc>, sender(m))
+  } else if quorumRound() > Rd {
+	Rd ← quorumRound()
+	Rc ← quorumRound()
+    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
+	if Vc != nil && isProposer(Hc, Rc) {
+      bc(<RC_T, Hc, Rc, Vc, PCc>)
+	}
+  } else if f1Round() > Rd {
+    Rd ← f1Round() 
+    Sc ← WaitingForNewRound
+    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
+    bc(<RC_T, Hc, Rd, PCc>)
+  }
+```
+
+### Functions
+
+#### Application provided functions
+No pseudocode is provided for these functions since their implementation is
+application specific.
+
+`proposer(H,R)`\
+Returns the proposer for the given height and round.
+
+`isProposer(H, R)`\
+Returns true if the current participant is the proposer for the given height
+and round. 
+
+`deliverValue(V)`\
+Delivers the given value to the application.
```

Comment:
What does this mean? What is 'deliver' here?

Reply:
https://github.com/celo-org/celo-blockchain/pull/1726#discussion_r738256951

Reply:
Ok, so it's a way of saying that this value is now accepted by the consensus? The name was misleading to me

Reply:
Not really, by the time `deliverValue` is called the value has already been committed by consensus. In other BFT consensus algorithms I have seen at that point they move to the next height and begin agreeing on the next value. But in this one the consensus algorithm passes the value to the application and waits for a notification from the application. `FinalComittedEvent` before continuing to the next height. I don't know why it does this.

In fact now that I'm thinking about it seems like this could be a source of problems in the protocol. If the final comitted event is delayed, then the network I think could move to the next round and agree on another value. Then you would have a fork.